packages/edit-site/src/components/header-edit-mode/index.js

```diff
@@ -259,6 +259,7 @@ export default function HeaderEditMode() {
 										/* translators: button label text should, if possible, be under 16 characters. */
 										label={ __( 'List View' ) }
 										onClick={ toggleListView }
+										ref={ setListViewToggleElement }
```

Comment:
Passing a `useState` callback seems fine to me. It's a stable reference. 

I was just wondering if it would cause some cognitive confusion for anyone in the future trying to access `current`. 

Another question might be: is `listViewToggleElement` (the state value)  a value that’s needed for rendering the component in which it lives, and would it trigger an unnecessary re-render if set?

This is all purely academic, and I'm only raising because I don't know myself! Something like this might be an alternative to `useState` in [layout](https://github.com/WordPress/gutenberg/pull/54175/files#diff-89e57bfd1de96b1cee155f91aa9754cb37715e9ffa9e4cc2f21463a770f7ec55R130) for example.

```js
	const listViewToggleElementRef = useRef( null );
	const setListViewToggleElementRef = useCallback( ( node ) => {
		if ( node !== null ) {
			listViewToggleElementRef.current = node;
		}
	}, [] );
```

Reply:
Thanks for taking a look @ramonjd, and good question! I mostly went with `useState` for consistency with some of the existing places where we're using ref callbacks to store elements in state (e.g. setDropZoneElement [here](https://github.com/WordPress/gutenberg/blob/3d2bffd7c0ddbbf4cd2ab1e927b416728bc0c7e9/packages/edit-post/src/components/secondary-sidebar/list-view-sidebar.js#L43)). In terms of trying to avoid confusion and whether or not `current` is available, I think so long as the variable is named with `Element` as its suffix instead of `Ref`, it'll hopefully be clear enough — or at least consistent enough with the other usage in the repo. Did you have any instances in mind where folks might reach for `current` and get tripped up?

In terms of re-renders, one of the benefits of going with state is that the consumer (`list-view-sidebar.js`) is fairly straightforward in that we can place `listViewToggleElement` in the dependency arrays of `useCallback`s and know that those callbacks will always be using the appropriate current value, without having to do any further introspection.

Some of the feedback I've gotten on previous PRs that looked at using refs with elements were to prefer using state... looking the references for that up now (if you'll excuse the pun 😄), I think the linked docs were in the floating UI library, so was more to do with working with popovers ([docs](https://floating-ui.com/docs/react#external-elements)), so might not be strictly speaking applicable here.

In any case, from the perspective of the list view sidebars, it feels a tiny bit neater to me if we're accessing the element directly in the component's props, rather than using a ref there, but happy to look into changing things if folks would prefer, especially if I've accidentally introduced re-renders in this PR where I shouldn't have!

Reply:
I appreciate the follow up explanation. Good to know there's a precedent. I've seen it in other libraries too, so TIL it's a thing for Gutenberg too. 🙇🏻